pkg/handler/collector/s3/s3.go

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/guacsec/guac/pkg/handler/collector/s3/bucket"
+	"github.com/guacsec/guac/pkg/handler/collector/s3/messaging"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+)
+
+const (
+	S3CollectorType = "S3CollectorType"
+)
+
+type S3Collector struct {
+	config S3CollectorConfig
+}
+
+type S3CollectorConfig struct {
+	MessageProvider     string
+	MessageProviderHost string
+	MessageProviderPort string
+	S3Host              string
+	S3Port              string
+	S3Bucket            string
+	S3Item              string
+	Region              string // optional (defaults to us-east-1, assumes same region for s3 and sqs)
+	Queues              string
+	MpBuilder           messaging.MessageProviderBuilder // optional
+	BucketBuilder       bucket.BuildBucket               // optional
+	SigChan             chan os.Signal                   // optional
+	Poll                bool
+}
+
+func NewS3Collector(cfg S3CollectorConfig) *S3Collector {
+	s3collector := &S3Collector{
+		config: cfg,
+	}
+	return s3collector
+}
+
+func (s S3Collector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+
+	if s.config.Poll {
+		retrieveWithPoll(s, ctx, docChannel)
+	} else {
+		retrieve(s, ctx, docChannel)
+	}
+
+	return nil
+}
+
+func retrieve(s S3Collector, ctx context.Context, docChannel chan<- *processor.Document) {
+	logger := logging.FromContext(ctx)
+	downloader := getDownloader(s)
+
+	bckt := s.config.S3Bucket
+	item := s.config.S3Item
+
+	blob, err := downloader.DownloadFile(ctx, bckt, item)
+	if err != nil {
+		logger.Errorf("could not download item %v: %v", item, err)
+		return
```

Comment:
This function should return this error, which RetrieveArtifacts should then return. Same with the one a couple lines down.